Region pinning preview docs (#1865)
ngrok/ngrok-docs: universal-gateway/region-pinning.mdx

description: Customize where your traffic connects to meet data residency needs, like keeping data within the EU.
tag: Preview
---

<Warning>
This feature is under active development. Behavior, supported fields, and limits may change before General Availability (GA). This guide is provided as forward-looking context for evaluation and feedback.

[Visit this page](https://dashboard.ngrok.com/developer-preview) to sign up for Early Access, or to send input ('Let us know' link at the bottom).
</Warning>

Region Pinning restricts which ngrok regions may handle traffic for a Domain. 

Control traffic into your [ngrok Domains](https://ngrok.com/docs/universal-gateway/domains) by choosing a subset of our globally distributed physical [Points of Presence](https://ngrok.com/docs/universal-gateway/points-of-presence). ngrok will only route to the set of regions you choose -- *not* outside it.

The same configuration will soon support [Dedicated IP addresses \\[Preview\\]](#dedicated-ips).

This feature is designed for customers with **data residency** and **compliance requirements** who need to ensure traffic is only processed in specific geographical boundaries.

# How it works:

> 1. Define which ngrok [Points of Presence](https://ngrok.com/docs/universal-gateway/points-of-presence) your [Domain](https://ngrok.com/docs/universal-gateway/domains) is allowed to use. 

Reviewer: Avoid capitalizing nouns that aren't proper nouns / ngrokisms. Also, let's introduce the acronym on first use and then use it from here on out. And avoid linking out to the same doc multiple times—once on first mention is enough.

```suggestion
1. Define which ngrok points of presence (PoPs) your domain is allowed to use. 
```

2. All requests to [Endpoints](https://ngrok.com/docs/universal-gateway/endpoints) on that Domain will ***only*** be routed through your chosen Points of Presence. 

<img 
  src="../img/docs/region_pinning_diagram_how_it_works.png" 
  alt="Region Pinning diagram -- how it works"
  style={{height: "auto", width: "50%"}}
/>

> # How to configure region pinning

Reviewer: ```suggestion
## How to configure region pinning
```

To start using region pinning: update your [Domain resource](https://ngrok.com/docs/universal-gateway/domains) with your preferred set of target regions.

> You can do this through the **API** or **dashboard**. 

Reviewer: ```suggestion
You can do this through the API or the dashboard, both described below. 
```

## Using the API

When using the API, you’ll setup Region Pinning by defining which [Points of Presence](https://ngrok.com/docs/universal-gateway/points-of-presence) (PoPs) or Region aliases should serve your requests. 

Just add the `ips` property when *creating* or *modifying* a Domain.

> ### Option 1: Create a pinned domain

Reviewer: ```suggestion
#### Option 1: Create a pinned domain
```

Use the `POST /reserved_domains` endpoint ([documentation](https://ngrok.com/docs/api-reference/reserveddomains/create)).

Add a list of `ips` that are allowed to resolve your requests: either a set of ngrok [physical Points of Presence](https://ngrok.com/docs/universal-gateway/points-of-presence) (like “Frankfurt” or “California”), or [Dedicated IPs \\[Preview\\]](#dedicated-ips), or both. 

```json
curl --request POST \
 --url https://api.ngrok.com/reserved_domains \
 --header "Authorization: Bearer <token>" \
 --header "Content-Type: application/json" \
 --header "ngrok-version: 2" \
 --data '{
    "domain": "api.example.com",
    "description": "Pinned to US and Europe for compliance reasons",
    "certificate_management_policy": {
      "authority": "letsencrypt",
      "private_key_type": "ecdsa"
    },
    "ips": [
      { "value": "us-ohio-1"},  
      { "value": "de-fra-1"}
    ]
  }'
```

### Option 2: Update an existing domain

You can update existing Domains to *add* Region Pinning, *remove* it, or *modify* your set of pinned Regions. 

> No DNS changes are required; just update the `ips` property in your Domain. 

Reviewer: ```suggestion
You can update existing domains to add or remove Region Pinning, or modify your set of pinned regions, using the `ips` property.
No DNS changes are required.
```

Use `PATCH` to update the list of pinned regions (see [API docs](https://ngrok.com/docs/api-reference/reserveddomains/update)).

```json
curl --request PATCH \
 --url https://api.ngrok.com/reserved_domains/{domain_id} \
 --header "Authorization: Bearer <token>" \
 --header "Content-Type: application/json" \
 --header "ngrok-version: 2" \
 --data '{
    "domain": "api.example.com",
    "description": "Pinned to US and Europe for compliance reasons",
    "certificate_management_policy": {
      "authority": "letsencrypt",
      "private_key_type": "ecdsa"
    },
    "ips": [
      { "value": "us-ohio-1"},  
      { "value": "de-fra-1"}
    ]
  }'
```

### Un-pin a domain

> You can “un-pin” a Domain, which changes it back to the default **Global** routing behavior (use the best-available PoP). 

Reviewer: ```suggestion
You can un-pin a domain, which changes it back to the default Global routing behavior using the best-available PoP. 
```

```json
# Only change the last property. The rest is same as above.

 --data '{
   "ips": [
     { "value": "global" }
   ]
 }'  
```

### API error handling

If a request has invalid syntax or a non-existent region code, the API request will fail.

The request returns **HTTP 400 Bad Request** with a structured error body, similar to this:

```json
HTTP/1.1 400 Bad Request
Content-Type: application/json; charset=utf-8

{
 "status_code": 400,
 "error_code": "invalid_region_target",
	 "msg": "Invalid region target 'banana': must be a valid Point of Presence or Regional Alias.",
 "details": {}
}
```

## Using the dashboard

You can also make these changes through the [ngrok dashboard](https://dashboard.ngrok.com/domains). 
<Warning>
*Preview Feature*. These designs are subject to change.
</Warning>

To configure your “pinned” regions: 

1. View your existing Domains at dashboard.ngrok.com/domains
2. Click into the Domain you wish to modify. 
3. Choose your target regions from the dropdown menu. 
    1. Select a set of [Points of Presence](https://ngrok.com/docs/universal-gateway/points-of-presence), or choose a “Region alias” that covers multiple PoPs, or both.

> ![Region Pinning dashboard](../img/docs/region-pinning-dashboard-ui.png)

Reviewer: As a general rule we try to avoid UI screenshots because 1) they're subject to change and 2) ideally the UI should be designed well enough that the user doesn't need us to point out elements on the screen! 😁

```suggestion
```

# Dedicated IPs
<Warning>
Preview feature, details subject to change. This is related to Region Pinning, but will be released separately. [Sign up for Early Access](https://dashboard.ngrok.com/developer-preview).
</Warning>

> Dedicated IPs let you configure a Domain’s matching endpoints to receive traffic on IP addresses that are **dedicated only to your account** — rather than using ngrok’s default range of multi-tenant IP addresses shared across our customers.

Reviewer: ```suggestion
Dedicated IPs let you configure a domain’s matching endpoints to receive traffic on IP addresses that are dedicated only to your account, rather than using ngrok’s default range of multi-tenant IP addresses shared across customers.
```

> When a Domain uses dedicated IPs, you can create matching endpoints on any inbound port (not just 80/443).

Reviewer: ```suggestion
When a domain uses dedicated IPs, you can create matching endpoints on any inbound port (not just 80/443).
```

### How to use it

> Dedicated IPs and Region Pinning share the same `ips` field on the Domain:

Reviewer: ```suggestion
Dedicated IPs and Region Pinning share the same `ips` field on the domain:
```

  ]
}

```

# FAQs (Frequently Asked Questions)

1. **Does this affect TCP endpoints?**
    - Region pinning: no. In this preview, region pinning applies to HTTP/HTTPS matching endpoints. TCP uses [reserved addresses](https://ngrok.com/docs/api-reference/reservedaddrs), which already include a region.
    - Dedicated IPs: yes, Dedicated IPs are available for *any* [ngrok-supported protocol](/universal-gateway/protocols):  HTTP, HTTPS, TCP, TLS
2. **What happens if I forget to add the `ips` field when creating a Domain?**
    1. A: it defaults back to standard **GLOBAL** routing, where traffic is automatically routed through the *best option* from our physical [Points of Presence](/universal-gateway/points-of-presence). This is the baseline behavior for Domains, and is used by the vast majority of customers. 
3. **Do I have to specify every region manually?** 
    1. No: we provide regional aliases as a “shortcut” to select *multiple* ngrok Points of Presence within a geographical area. 
    2. Currently, the only supported alias is `global`. 
    3. *We may add other aliases like `european_union` or `united_states` , depending on customer needs.*

# Reference: Points of Presence

> The `ips` parameter uses these new, clearer PoP names:

Reviewer: ```suggestion
The `ips` parameter uses the following PoP names:
```
